llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -6505,6 +6505,33 @@ Value *llvm::simplifyBinaryIntrinsic(Intrinsic::ID IID, Type *ReturnType,
 
     break;
   }
+  case Intrinsic::scmp:
+  case Intrinsic::ucmp: {
+    // Fold cmp x, x -> 0
+    if (Op0 == Op1)
+      return Constant::getNullValue(ReturnType);
+
+    // Fold to a constant if the relationship between operands can be
+    // established with certainty
+    if (isICmpTrue(CmpInst::ICMP_EQ, Op0, Op1, Q, RecursionLimit))
```

Comment:
Do we need `Q.getWithoutUndef()` here?

Reply:
I don't think so, why?

Reply:
See https://github.com/llvm/llvm-project/blob/355e4a9e56c644f24fc10f780cb2fc68b660d0a0/llvm/lib/Analysis/InstructionSimplify.cpp#L6498-L6503

There is no negative test which blocks undef fold in https://reviews.llvm.org/D85929.

Reply:
The situation with min/max is different. If you have something like `umin(undef, 5)` then `icmp ule undef, 5` could fold to true and return `undef` overall, which would be incorrect. In this case we return a constant, which is fine (in the same way that folding an icmp with undef is fine -- if this is not valid for the using context, the caller would already disable undef refinement).

Reply:
It makes sense to me. Thank you!